platformio.ini

```diff
@@ -40,6 +40,16 @@ build_flags = ${env.build_flags} -DBARE_STM32 -mfpu=fpv4-sp-d16 -mfloat-abi=hard
 board_build.ldscript = boards/stm32_ldscript.ld
 build_unflags = -std=gnu11 -std=gnu++14
 extra_scripts = boards/stm32_scripts.py
+src_filter = ${env.src_filter} -<test/> -<src_test/>
+
+[env:stm32-test]
+platform = ststm32
+board = custom_stm32
+build_flags = ${env.build_flags} -Wconversion -Wno-sign-conversion -Wno-error=register -mfpu=fpv4-sp-d16 -mfloat-abi=hard -DBARE_STM32 -DUART_VIA_DMA -Wl,-Map,stm32.map -Wl,-u,vectors -Wl,-u,_init
+board_build.ldscript = boards/stm32_ldscript.ld
+build_unflags = -std=gnu11 -std=gnu++14
+extra_scripts = boards/stm32_scripts.py
+src_filter = ${env.src_filter} -<test/> -<src/> +<src_test/>
```

Comment:
I don't think you should add a UART handler that isn't actually being used.  If you're going to add this, you should enable it and replace the older UART code so it gets exercised in the system.  We don't want a lot of dead code that's never really been used lying around.

Reply:
Replacing all UART code with this implementation has bigger implications on `comms.cpp` I wanted to avoid. I wanted to decouple this driver rollout from the rest of 159 related stuff that's now happening in #333 reason being that this PR is big enough itself. #333 already has previous main UART code removed and replaced with this UART_DMA implementation, debug UART coming along.

Regarding `strm32-test` environment - this is to provide reviewers a chance to see the code in action, also an illustration what integration testing might look like. I am happy to remove it in case you think we don't want that in master branch.

Reply:
Yes, I share your concerns.  I know that this UART driver will completely break the debug interface that I'm trying to add in my issue 301 branch which makes me sad.

Are you sure we really benefit from a DMA based UART at all?  The serial port is relatively slow.  At 115.2k it will send a character not more frequently then once / 87 microseconds.  That won't really have any noticable impact on our processor's work load.  

Also, due to framing the serial messages that we receive are of variable data length.  Actually, that's true for my debug interface I really haven't spent much time looking at the other serial interface.  DMA works best if you know the size of the message and can just set the DMA up to handle it and then forget about it.  I don't think that works so well in our case.

In reviewing this code I'm really just looking for obvious bugs or small improvements that can be made to the implementation.  Other then the few minor issues I pointed out I don't see any major problems with this driver that would stop me from approving it.

From a systems architecture standpoint, I'm am not convinced that it's the right solution.  I'm not really sure if anyone is keeping an eye on overall system architecture though and I don't think that can be done piecemeal in the PR reviews.  It's something we should all discuss at the next software meeting.

I'm going to approve the PR because I don't see any major problems to the code.  Perhaps you're right though, it's probably better not to enable this driver just yet.

Reply:
My opinion is that UART via DMA is beneficial to this project. When txing we work on buffers of a known size, that was introduced by the decision made to use protobuf. Protobuf allows us not to do message coding algorithmically, enabling us to easily change the message structure, which will happen a lot when the project evolves. There is also a stream object in nanopb library, but provided we produce streams that are bigger (+crc +esc chars) than the input stream, we would need to have buffered streams and that sounds kinda complicated already and denies the nice idea of a lean simple stream, so no stream paradigm. So a buffer of a known size containing an encoded message it is. When I was working on a low end avr and pics, I was longing for a peripheral that would just bang my buffer via uart no questions asked. Well, on this MCU, DMA is such peripheral. 

Regarding rxing - we need to assemble the whole message before we can give it to protobuf. Start and end of the message is marked with a symbol, and we luckily have char match interrupt, so with a small, testable FSM and DMA we get full message in the buffer ready for us to decode.

Frankly, for me *this* sounds less complicated than having to do it by hand :) But I was lucky with DMA in projects before - it was smooth and snappy, so maybe this is just that.

Thank you for your review.